docs/design/cloud-init.rst

```diff
@@ -0,0 +1,260 @@
+CLOUD_INIT image property.
+
+Proposed changes
+================
+
+Datasources
+-----------
+
+Since we want snf-image to be able to work stand-alone, without having to
+depend of external services, the software could make use of only 2 datasources:
+* NoCloud
+* None
+All the others expect to fetch data from an external entity (CDROM, Floppy,
+serial-port, Web-service).
+
+The "None" is the fallback datasource when no other can be selected. It is the
+equivalent of an empty datasource in that it provides an empty string as
+userdata and an empty dictionary as metadata. We could make use of this and let
+*snf-image-helper* inject all it's configuration to files under
+``/etc/cloud/cloud.cfg.d``. We could then allow the user to provide extra
+configuration data by using an new OS parameter (e.g. cloud_usedata) which's
+content would be injected into a file under ``/etc/cloud/cloud.cfg.d``. The
+only problem with this approach is that the user may only provide YAML user
+data, which is not fully compatible with Openstack. The nova API allows the
+user to provide user-data upon server creation. Those user data are available
+in the OpenStack meta-data service to be consumed by cloud-init. Since we are
+using *snf-image* to build a stack that exposes an OpenStack compute API, it's
+highly discouraged to have a, by design, incompatibility.
+
+This leaves us with the NoCloud option. This datasource allows us to provide
+user-data and meta-data to the instance through a number of different ways. The
+most suitable for *snf-image* is the ``/var/lib/cloud/seed/nocloud-net``
+directory. The datasource will expect to find the files ``meta-data`` and
+``user-data`` and optionally ``verdor-data`` as well as ``network-config``
+under this directory. *snf-image* could use a dedicated task to created those
+files and write this:
+```
+datasource_list [NoCloud]
+```
+to the cloud-init configuration. The various tasks could then perform their
+configuration by truncating the ``meta-data`` and ``vendor-data`` files or by
+appending files to the cloud-init configuration directory. By definition, the
+best place for *snf-image* to put the configuration would be the vendor-data:
+
+> Vendordata is data provided by the entity that launches an instance (for
+> example, the cloud provider). This data can be used to customize the image to
+> fit into the particular environment it is being run in.
+
+Unfortunately, older versions of cloud-init don't play well with
+``vendor-data``, which leaves us with ``meta-data`` and the files under
+``/etc/cloud/cloud.cfg.d``.
+
+Configuration Tasks
+-------------------
+
+*snf-image* configures the VM by running a number of configuration tasks on the
+hard disk of the VM. On images that support cloud-init, instead of directly
+altering the needed system files, *snf-image-helper* could add cloud-init
+configuration into the meta-data file as well as files under
+``/etc/cloud/cloud.cdg.d/``.
```

Comment:
There isn't something special about the cloud-init config module logic. Each cloud-init config module in a separate python module. They are all hosted under the ``cloud-init/config`` directory and cloud-init expects them to implement the function ``def handle(name, cfg, cloud, log, _args)``. Each module does something based on some configuration options. Those configuration options are not namespaced based on the module or something. The user does not need to know anything about the cloud-init config modules. They are mentions here because the configuration options in the cloud-init documentation are ordered by the module they are used from.